core/drivers/versal_trng.c

```diff
@@ -0,0 +1,1147 @@
+	};
+	const uint8_t expected_out[TRNG_GEN_LEN] = {
+		0x91U, 0x9AU, 0x6BU, 0x99U, 0xD5U, 0xBCU, 0x2CU, 0x11U,
+		0x5FU, 0x3AU, 0xFCU, 0x0BU, 0x0EU, 0x7BU, 0xC7U, 0x69U,
+		0x4DU, 0xE1U, 0xE5U, 0xFEU, 0x59U, 0x9EU, 0xAAU, 0x41U,
+		0xD3U, 0x48U, 0xFDU, 0x3DU, 0xD2U, 0xC4U, 0x50U, 0x1EU,
+	};
+	uint8_t out[TRNG_GEN_LEN] = { 0 };
+
+	if (!trng)
+		return TEE_ERROR_GENERIC;
+
+	memcpy(&tests.init_seed, ext_seed, sizeof(ext_seed));
+	memcpy(tests.pstr, pers_str, sizeof(pers_str));
+
+	if (trng_instantiate(trng, &tests))
+		goto error;
+
+	if (trng_generate(trng, out, sizeof(out), false))
+		goto error;
+
+	if (memcmp(out, expected_out, TRNG_GEN_LEN)) {
+		EMSG("K.A.T mismatch");
+		goto error;
+	}
+
+	if (trng_release(trng))
+		goto error;
+
+	return TEE_SUCCESS;
+error:
+	trng->status = TRNG_ERROR;
+	return TEE_ERROR_GENERIC;
+}
+
+static struct versal_trng versal_trng = {
+	.cfg.base = TRNG_BASE,
+	.cfg.len = TRNG_SIZE,
+};
+
+TEE_Result hw_get_random_bytes(void *buf, size_t len)
+{
+	uint8_t random[TRNG_SEC_STRENGTH_LEN] = { 0 };
+	uint8_t *p = buf;
+	size_t i = 0;
+
+	for (i = 0; i < len / TRNG_SEC_STRENGTH_LEN; i++) {
+		if (trng_generate(&versal_trng, p + i * TRNG_SEC_STRENGTH_LEN,
+				  TRNG_SEC_STRENGTH_LEN, false))
+			panic();
+	}
+
+	if (len % TRNG_SEC_STRENGTH_LEN) {
+		if (trng_generate(&versal_trng, random, TRNG_SEC_STRENGTH_LEN,
+				  false))
+			panic();
+		memcpy(p + i * TRNG_SEC_STRENGTH_LEN, random,
+		       len % TRNG_SEC_STRENGTH_LEN);
+	}
+
+	return TEE_SUCCESS;
+}
+
+void plat_rng_init(void)
+{
+}
+
+static TEE_Result trng_hrng_mode_init(void)
+{
+	const uint8_t pers_str[TRNG_PERS_STR_LEN] = {
+		0xB2, 0x80, 0x7E, 0x4C, 0xD0, 0xE4, 0xE2, 0xA9,
+		0x2F, 0x1F, 0x5D, 0xC1, 0xA2, 0x1F, 0x40, 0xFC,
+		0x1F, 0x24, 0x5D, 0x42, 0x61, 0x80, 0xE6, 0xE9,
+		0x71, 0x05, 0x17, 0x5B, 0xAF, 0x70, 0x30, 0x18,
+		0xBC, 0x23, 0x18, 0x15, 0xCB, 0xB8, 0xA6, 0x3E,
+		0x83, 0xB8, 0x4A, 0xFE, 0x38, 0xFC, 0x25, 0x87,
+	};
+	/* configure in hybrid mode with derivative function enabled */
+	struct trng_usr_cfg usr_cfg = {
+		.mode = TRNG_HRNG,
+		.seed_life = CFG_VERSAL_TRNG_SEED_LIFE,
+		.predict_en = false,
+		.df_disable = false,
+		.dfmul = CFG_VERSAL_TRNG_DF_MUL,
+		.iseed_en =  false,
+		.pstr_en = true,
+	};
+
+	memcpy(usr_cfg.pstr, pers_str, TRNG_PERS_STR_LEN);
+	versal_trng.cfg.addr = (vaddr_t)core_mmu_add_mapping(MEM_AREA_IO_SEC,
+							versal_trng.cfg.base,
+							versal_trng.cfg.len);
+	if (!versal_trng.cfg.addr) {
+		EMSG("Failed to map TRNG");
+		goto error;
+	}
+
+	if (trng_kat_test(&versal_trng)) {
+		EMSG("KAT Failed");
+		goto error;
+	}
+
+	if (trng_health_test(&versal_trng)) {
+		EMSG("RunHealthTest Failed");
+		goto error;
+	}
+
+	if (trng_instantiate(&versal_trng, &usr_cfg)) {
+		EMSG("Driver instantiation Failed");
+		goto error;
+	}
+
+	if (trng_reseed(&versal_trng, NULL, usr_cfg.dfmul)) {
+		EMSG("Reseed Failed");
+		goto error;
+	}
+
+	return TEE_SUCCESS;
+error:
+	panic();
+	return TEE_ERROR_GENERIC;
```

Comment:
Remove the return instruction.

Maybe replace `goto error` with `panic` straight? Whatever, do as you prefer.